src/libs/sfdk/dockervirtualmachine.cpp

```diff
@@ -0,0 +1,494 @@
+{
+    QJsonDocument document = QJsonDocument::fromJson(output);
+    VirtualMachineInfo info = {};
+
+    if (document.isNull())
+        return info;
+
+    QJsonObject labels = document.object();
+
+    auto parsePort = [=](VirtualMachinePrivate::ReservedPort which, quint16& port, quint16 guestPort) {
+        QString key = label(which);
+        if (labels.contains(key)) {
+            port = labels.value(key).toInt();
+            m_reservedPortMappings[which] = QPair<quint16, quint16>(port, guestPort);
+        }
+    };
+
+    parsePort(VirtualMachinePrivate::SshPort, info.sshPort, GUESTSSHPORT);
+    parsePort(VirtualMachinePrivate::WwwPort, info.wwwPort, GUESTWWWPORT);
+
+    auto parsePath = [=](VirtualMachinePrivate::SharedPath which, QString& path, const QString& guestPath) {
+        QString key = label(which);
+        if (labels.contains(key)) {
+            path = labels.value(key).toInt();
+            m_sharedPaths[which] = QPair<QString, QString>(path, guestPath);
+        }
+    };
+
+    parsePath(VirtualMachinePrivate::SharedSrc, info.sharedSrc, Constants::BUILD_ENGINE_GUEST_TARGET_PATH);
+    parsePath(VirtualMachinePrivate::SharedSsh, info.sharedSsh, Constants::BUILD_ENGINE_SHARED_SSH);
+    parsePath(VirtualMachinePrivate::SharedHome, info.sharedHome, Constants::BUILD_ENGINE_SHARED_HOME);
+    parsePath(VirtualMachinePrivate::SharedConfig, info.sharedConfig, Constants::BUILD_ENGINE_SHARED_CONFIG);
+    parsePath(VirtualMachinePrivate::SharedTargets, info.sharedTargets, Constants::BUILD_ENGINE_SHARED_TARGET);
+    return info;
+}
+
+
+void DockerVirtualMachinePrivate::fetchInfo(VirtualMachineInfo::ExtraInfos extraInfo,
+    const QObject *context, const Functor<const VirtualMachineInfo &, bool> &functor) const
+{
+    Q_Q(const DockerVirtualMachine);
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+
+    // TODO: fetch snapshot info if requested
+    Q_UNUSED(extraInfo)
+
+    QStringList arguments;
+    arguments.append(COMMAND_INSPECT);
+    arguments.append(q->name());
+    arguments.append(FORMAT_LABEL_LIST);
+
+    auto info = std::make_shared<VirtualMachineInfo>();
+
+    auto runner = std::make_unique<ProcessRunner>(DockerManager::dockerPath(), arguments);
+    QObject::connect(runner.get(), &ProcessRunner::done, q, [this, functor, info, process = runner->process()](bool ok) {
+        if (ok)
+            *info = virtualMachineInfoFromOutput(process->readAllStandardOutput().trimmed());
+
+        functor(*info, ok);
+    });
+    SdkPrivate::commandQueue()->enqueue(std::move(runner));
+    SdkPrivate::commandQueue()->enqueueCheckPoint(context, [=]() { functor(*info, true); });
+}
+
+void DockerVirtualMachinePrivate::start(const QObject *context, const Functor<bool> &functor)
+{
+    Q_ASSERT(context);
+    Q_ASSERT(functor);
+    Q_Q(DockerVirtualMachine);
+
+    CommandQueue::BatchId batch = SdkPrivate::commandQueue()->beginBatch();
+
+    if (m_reservedPortMappings.size() == 0) {
```

Comment:
I think we should assume (and ensure*) that `start` is never called before the info is fetched.

*) If it happened for you, then `enableUpdates` and `updateOnce` under `BuildEnginePrivate` and `EmulatorPrivate` may need to be fixed, delaying the `setAutoConnectEnabled` call until the `updateVmProperties` call finishes.

Reply:
I'll take your word for it - and add an assert.